src/flexalloc_xnvme_env.c

```diff
@@ -189,102 +173,195 @@ fla_xne_sync_seq_w_naddrs(struct xnvme_dev * dev, const uint64_t slba, const uin
+  struct fla_async_strp_cb_args *cb_arg;
+  struct xnvme_cmd_ctx *ctx;
+  uint32_t nsid = xnvme_dev_get_nsid(dev);
 
-  err = xnvme_queue_init(dev, strp_nobjs, 0, &queue);
+  struct fla_async_strp_cb_args cb_args [512] = {0};
+
+  err = xnvme_queue_init(dev, sp->strp_nobjs * 2, 0, &queue);
   if (FLA_ERR(err, "xnvme_queue_init"))
     goto exit;
 
-  xnvme_queue_set_cb(queue, fla_async_cb, &cb_args);
-
-  while (bytes_xfer != nbytes)
+  uint64_t sbuf_nbytes = 0, xfer_snbytes = sp->xfer_snbytes;
+  for(uint32_t i = 0 ; i < sp->strp_nobjs && sbuf_nbytes < sp->xfer_nbytes; ++i)
   {
+    ctx = xnvme_queue_get_cmd_ctx(queue);
+    uint16_t curr_nlbs = calc_strp_obj_first_nlbs(xfer_snbytes, sp, geo->lba_nbytes);
+    cb_arg = &cb_args[i];
+    cb_arg->nsid = nsid;
+    cb_arg->buf = (char *)buf;
+    cb_arg->nlbs = curr_nlbs;
+    cb_arg->slba = calc_strp_obj_slba(xfer_snbytes, sp, geo->lba_nbytes)
+                   + (sp->strp_obj_start_nbytes/geo->lba_nbytes);
+    cb_arg->sp = sp;
+    cb_arg->dev_lba_nbytes = geo->lba_nbytes;
+    cb_arg->sbuf_nbytes = sbuf_nbytes;
+    cb_arg->ebuf_nbytes = sp->xfer_nbytes;
+    cb_arg->write = write;
+
+    xnvme_cmd_ctx_set_cb(ctx, fla_async_strp_cb, cb_arg);
 
-    for (uint32_t strp=0; strp < strp_nobjs;)
-    {
-      struct xnvme_cmd_ctx *ctx = xnvme_queue_get_cmd_ctx(queue);
-      uint64_t slba = (strp_offset/geo->lba_nbytes) + (strp * obj_nlbs);
 submit:
-      if (write)
-        err = xnvme_nvm_write(ctx, nsid, slba, strp_blks, strp_buf + (strp * strp_nbytes), NULL);
-      else
-        err = xnvme_nvm_read(ctx, nsid, slba, strp_blks, strp_buf + (strp * strp_nbytes), NULL);
-
-      switch (err)
-      {
-      case 0:
-        cb_args.submitted +=1;
-        goto next;
-
-      case -EBUSY:
-      case -EAGAIN:
-        xnvme_queue_poke(queue, 0);
-        goto submit;
-
-      default:
-        FLA_ERR(1, "Async submission error\n");
-        goto exit;
-      }
-
-next:
-      ++strp;
-    }
+    err = write
+          ? xnvme_nvm_write(ctx, cb_arg->nsid, cb_arg->slba, cb_arg->nlbs,
+                            cb_arg->buf + cb_arg->sbuf_nbytes, NULL)
+          : xnvme_nvm_read(ctx, cb_arg->nsid, cb_arg->slba, cb_arg->nlbs,
+                           cb_arg->buf + cb_arg->sbuf_nbytes, NULL);
 
-    err = xnvme_queue_wait(queue);
-    bytes_xfer += strp_nbytes * strp_nobjs;
-    strp_buf += strp_nbytes * strp_nobjs;
-    strp_offset += strp_nbytes;
-    if (FLA_ERR(cb_args.ecount, "Error completing async IO\n"))
-      goto exit;
+    switch (err)
+    {
+    case 0:
+      cb_arg->cb_args.submitted +=1;
+      break;
+
+    case -EBUSY:
+    case -EAGAIN:
+      xnvme_queue_poke(queue, 0);
+      goto submit;
+
+    default:
+      FLA_ERR(1, "Async submission error\n");
+      goto close_queue;
+    }
 
+    sbuf_nbytes += (curr_nlbs + 1) * geo->lba_nbytes;
+    xfer_snbytes += (curr_nlbs + 1) * geo->lba_nbytes;
   }
 
-exit:
+close_queue:
+  ret = xnvme_queue_wait(queue);
```

Comment:
Can you verify with xnvme team that queue_wait will not return unless the cb is executed? I just want to make sure we don't have any conditions where we return from queue_wait but we don't generate all of the IO we expect.